src/ast/ast_build_op_contexts.c

```diff
@@ -0,0 +1,343 @@
+
+        // Updated value
+        const cypher_astnode_t *val_to_set = cypher_ast_set_property_get_expression(set_item); // type == CYPHER_AST_SET_PROPERTY
+
+        /* Track all required information to perform an update. */
+        update_expressions[i].attribute = cypher_ast_prop_name_get_value(prop);
+        update_expressions[i].exp = AR_EXP_FromExpression(record_map, val_to_set);
+        update_expressions[i].entityRecIdx = RecordMap_FindOrAddAlias(record_map, entity->operand.variadic.entity_alias);
+
+        AR_EXP_Free(entity);
+    }
+
+    *nitems_ref = nitems;
+    return update_expressions;
+}
+
+void AST_PrepareDeleteOp(const cypher_astnode_t *delete_clause, const QueryGraph *qg, RecordMap *record_map, uint **nodes_ref, uint **edges_ref) {
+    uint delete_count = cypher_ast_delete_nexpressions(delete_clause);
+    uint *nodes_to_delete = array_new(uint, delete_count);
+    uint *edges_to_delete = array_new(uint, delete_count);
+
+    for (uint i = 0; i < delete_count; i ++) {
+        const cypher_astnode_t *ast_expr = cypher_ast_delete_get_expression(delete_clause, i);
+        assert(cypher_astnode_type(ast_expr) == CYPHER_AST_IDENTIFIER);
+        const char *alias = cypher_ast_identifier_get_name(ast_expr);
+
+        // Ensure the entity is mapped in the Record
+        uint id = RecordMap_FindOrAddAlias(record_map, alias);
+
+        /* We need to determine whether each alias refers to a node or edge.
+         * Currently, we'll do this by consulting with the QueryGraph. */
+        SchemaType type = QueryGraph_GetEntityTypeByAlias(qg, alias);
+        if (type == SCHEMA_NODE) {
+            nodes_to_delete = array_append(nodes_to_delete, id);
+        } else if (type == SCHEMA_EDGE) {
+            edges_to_delete = array_append(edges_to_delete, id);
+        } else {
+            assert(false);
+        }
+    }
+
+    *nodes_ref = nodes_to_delete;
+    *edges_ref = edges_to_delete;
+
+}
+
+// TODO largely unnecessary
+int AST_PrepareSortOp(const cypher_astnode_t *order_clause) {
+    assert(order_clause);
+
+    bool ascending = true;
+    unsigned int nitems = cypher_ast_order_by_nitems(order_clause);
+
+    for (unsigned int i = 0; i < nitems; i ++) {
+        const cypher_astnode_t *item = cypher_ast_order_by_get_item(order_clause, i);
+        // TODO direction should be specifiable per order entity
+        ascending = cypher_ast_sort_item_is_ascending(item);
+    }
+
+    int direction = ascending ? DIR_ASC : DIR_DESC;
+
+    return direction;
+}
+
+AST_UnwindContext AST_PrepareUnwindOp(const cypher_astnode_t *unwind_clause, RecordMap *record_map) {
+    const cypher_astnode_t *collection = cypher_ast_unwind_get_expression(unwind_clause);
+    AR_ExpNode **exps = _AST_ConvertCollection(collection, record_map);
+    const char *alias = cypher_ast_identifier_get_name(cypher_ast_unwind_get_alias(unwind_clause));
+    uint record_idx = RecordMap_LookupAlias(record_map, alias);
```

Comment:
`UNWIND [1,2,3] AS X`
What guarantee do you have that `X` is in `record_map` ?

Reply:
:+1: Good catch, it's not guaranteed unless X is also returned. I'll fix that.